thunder/dynamo/splitter_experiment.py

```diff
@@ -0,0 +1,210 @@
+# CapabilityBasedParitioner returns a GraphModule where `fused_*` represent the subgraphs
+# that should go to `thunder` and the forward of this graph module should be passed to `torch.compile` (after removing thunder bits)
+# Example -
+# class GraphModule(torch.nn.Module):
+#     def forward(self, L_x_: "f32[2]"):
+#         l_x_ = L_x_
+
+#         # No stacktrace found for following nodes
+#         _enter_autocast = torch.amp.autocast_mode._enter_autocast('cpu', None, True, None)
+
+#          # File: /home/kkalambarkar/lightning-thunder/scratchpad/test.py:181 in func, code: return torch.matmul(x, y)
+#         fused_0: "bf16[]" = self.fused_0(l_x_);  l_x_ = None
+
+#         # No stacktrace found for following nodes
+#         _exit_autocast = torch.amp.autocast_mode._exit_autocast(_enter_autocast);  _enter_autocast = _exit_autocast = None
+#         return (fused_0,)
+
+#     class fused_0(torch.nn.Module):
+#         def forward(self, l_x_: "f32[2]"):
+#              # File: /home/kkalambarkar/lightning-thunder/scratchpad/test.py:177 in func, code: x = x + 2
+#             x: "f32[2]" = l_x_ + 2;  l_x_ = None
+
+#              # File: /home/kkalambarkar/lightning-thunder/scratchpad/test.py:179 in func, code: z = torch.ones(3, 3)
+#             z: "f32[3, 3]" = torch.ones(3, 3);  z = None
+
+#              # File: /home/kkalambarkar/lightning-thunder/scratchpad/test.py:180 in func, code: y = torch.sin(x)
+#             y: "f32[2]" = torch.sin(x)
+
+#              # File: /home/kkalambarkar/lightning-thunder/scratchpad/test.py:181 in func, code: return torch.matmul(x, y)
+#             matmul: "bf16[]" = torch.matmul(x, y);  x = y = None
+#             return matmul
+
+
+def capability_partitioner_splitter(gm, sample_args):
+    gm_copy = copy.deepcopy(gm)
+    op_support = ThunderOperatorSupport(gm_copy)
+    partitioner = CapabilityBasedPartitioner(gm_copy, op_support)
```

Comment:
How do you like their design of instantiating a separate class for operator support? I don't like and here's how we can avoid using it (by implementing `__is_node_supported`):
```py
class ThunderPartitioner(CapabilityBasedPartitioner):
    def __init__(self, gm):
        supported_ops = None # This is overridden by __is_node_supported
        super().__init__(gm, supported_ops, allows_single_node_partition=True)

    def __is_node_supported(self, node: torch.fx.Node) -> bool:
        # node.meta could be inspected for metadata about the node
        # For now, we just allow all nodes. Note that this is not a good idea
        # and ThunderSupport should be updated to only allow nodes that are
        # supported by Thunder
        return True
```

Reply:
With this subclass we can also override `fuse_partitions` to support passing our own name:
```py
class ThunderPartitioner(CapabilityBasedPartitioner):
    ...
    def fuse_partitions(self, partitions: list[Partition]) -> torch.fx.GraphModule:
        # fuse_by_partitions expects partitions in list[List[Node]]: [ [node0, node1], [node2, node3] ]
        return fuse_by_partitions(self.graph_module, [list(partition.nodes) for partition in partitions])

# fx.passes.utils.fuser_utils.fuse_by_partitions but with the name_prefix parameter
def fuse_by_partitions(gm: torch.fx.GraphModule, partitions: list[torch.fx.Node], name_prefix: str) -> torch.fx.GraphModule:
    from torch.fx.passes.utils.fuser_utils import topo_sort, insert_subgm, erase_nodes, fuse_as_graphmodule, legalize_graph
    for partition_id, nodes in enumerate(partitions):
        sorted_nodes = topo_sort(nodes)

        submodule_name = name_prefix + str(partition_id)
        sub_gm, orig_inputs, orig_outputs = fuse_as_graphmodule(gm, sorted_nodes, submodule_name)

        insert_subgm(gm, sub_gm, orig_inputs, orig_outputs)
        erase_nodes(gm, sorted_nodes)

    legalize_graph(gm)
    return gm
```

Reply:
Have removed the usage of `OperatorSupport` class and have function instead. 

Also, the function only cares about whether a node is supported by thunder and the logic whether it is in unsupported ctx region has been moved outside.